feat: implementa componente Input

ICarlosMarinho opened the pull request:
Implementa versão inicial do componente, bem como estilos e testes.

ICarlosMarinho committed:
feat: implementa componente Input

Implementa versão inicial do componente, bem como estilos e testes.

ICarlosMarinho commented on src/components/Input/index.tsx:
```diff
@@ -0,0 +1,66 @@
+import "./style.css";
+
+type InputConstraint = {
+  validator: (value: string) => boolean;
```
Essa função recebe o valor preenchido no input e retorna um _boolean_ baseado nesse valor. No caso deve retornar _false_, caso o valor preenchido seja inválido.

ICarlosMarinho commented on src/components/Input/index.tsx:
```diff
@@ -0,0 +1,66 @@
+
+  const handleInputBlur = () => {
+    if (constraints) {
+      const constraint = constraints.find(({ validator }) => validator(value) === false);
```
A ideia é receber um array de InputConstraint e exibir sempre o primeiro erro encontrado. Nesse caso só é exibido um erro por vez.

michellehorn replied:
vale a pena testar esse cenário

ICarlosMarinho commented on src/components/Input/test.tsx:
```diff
@@ -0,0 +1,65 @@
+import { screen, render, fireEvent } from "@testing-library/react";
+import Input from ".";
+
+describe("Tests Input component", () => {
```
Será que estou deixando de testar algo importante?

michellehorn replied:
```suggestion
describe("Input | Unit test , () => {
```

michellehorn commented on src/components/Input/test.tsx:
```diff
@@ -0,0 +1,65 @@
+
+    fireEvent.change(input, { target: { value } });
+
+    expect(onValueChange).toHaveBeenCalledWith(value);
```
🥳 nice

michellehorn commented on src/components/Input/test.tsx:
```diff
@@ -0,0 +1,65 @@
+    const input = screen.getByLabelText(labelText);
+
+    fireEvent.blur(input);
+    screen.getByText(message);
```
faltou o expect

ICarlosMarinho committed:
Update src/components/Input/test.tsx

Co-authored-by: Michelle Horn <michellehorn@live.com>

ICarlosMarinho committed:
config: adiciona extend-expect ao ambiente de testes

ICarlosMarinho committed:
refactor: adiciona testes

Adiciona teste que valida se o erro não é exibido caso a entrada do usuário seja válida.

Refatora teste que valida se o erro é renderizado, de forma que além dessa condição, o teste
valida também se apenas um erro é exibido por vez.
